src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -376,34 +338,43 @@ private static object BindToCollection(Type type, IConfiguration config, BinderO
                         return;
                     }
 
-                    object? boundFromInterface = AttemptBindToCollectionInterfaces(type, config, options);
-                    if (boundFromInterface != null)
+                    // For other mutable interfaces like ICollection<> and ISet<>, we prefer copying values and setting them
+                    // on a new instance of the interface over populating the existing instance implementing the interface.
+                    // This has already been done, so there's not need to check again. For dictionaries, we fill the existing
+                    // instance if there is one (which hasn't happened yet), and only create a new instance if necessary.
+                    if (TypeIsADictionaryInterface(type))
                     {
-                        bindingPoint.SetValue(boundFromInterface);
-                        return; // We are already done if binding to a new collection instance worked
+                        Type typeOfKey = type.GenericTypeArguments[0];
+                        Type typeOfValue = type.GenericTypeArguments[1];
+                        // Overwrite type in case it was a IReadOnlyDictionary<>. We still want to be able to bind items.
+                        // REVIEW: What about settable IReadOnlyDictionary<> instances with an initial value?
+                        // I think we should consider preferring copying like we do for all other collection interfaces.
+                        type = typeof(Dictionary<,>).MakeGenericType(typeOfKey, typeOfValue);
                     }
 
                     bindingPoint.SetValue(CreateInstance(type, config, options));
                 }
 
-                // See if it's a Dictionary
-                Type? collectionInterface = FindOpenGenericInterface(typeof(IDictionary<,>), type);
-                if (collectionInterface != null)
+                // At this point we know that we have a non-null bindingPoint.Value, we just have to populate the items
+                // using the IDictionary<> or ICollection<> interfaces, or properties using reflection.
+                Type? dictionaryInterface = FindOpenGenericInterface(typeof(IDictionary<,>), type);
+                Type? dictionaryInterface2 = FindOpenGenericInterface(typeof(IReadOnlyDictionary<,>), type);
```

Comment:
1. `dictionaryInterface2` doesn't seem like a great variable name.
2. Do we need to check for `IReadOnlyDictionary` unconditionally, even if we already found `IDictionary`?

Reply:
Removed the latter check